src/main/java/com/aliyun/odps/jdbc/OdpsPreparedStatement.java

```diff
@@ -81,6 +133,170 @@ public void clearParameters() throws SQLException {
   }
 
   @Override
+  public void clearBatch() throws SQLException {
+    batchedRows.clear();
+  }
+
+  /**
+   * Only support DML like `INSERT INTO table_name values (e, f, g)` in batch execution
+   *
+   * Since ODPS SQL does not provide this functionality, we 1) hijack such kind of batched SQLs
+   * , 2) assemble the records by ourselves, and 3) call tunnel API to upload them.
+   *
+   * We verify and parse the SQL to extract table name in the first call to executeBatch(), so that
+   * other kinds of statement can be executed in a non-batch way.
+   *
+   * Considering performance issue, We check it lazily in executeBatch() instead of addBatch().
+   *
+   * @throws SQLException
+   */
+  @Override
+  public int[] executeBatch() throws SQLException {
+
+    if (!verified) {
+      if (!sql.matches(PREP_INSERT_WITHOUT_SPEC_COLS)) {
+        throw new SQLException("batched statement only support following syntax: " +
+                               PREP_INSERT_WITHOUT_SPEC_COLS_EXAMPLE);
+      }
+
+      Matcher matcher = Pattern.compile(PREP_INSERT_WITHOUT_SPEC_COLS).matcher(sql);
+      if (matcher.find()) {
+        tableBatchInsertTo = matcher.group(1);
+      } else {
+        throw new SQLException("cannot extract table name in SQL: " + sql);
+      }
+
+      verified = true;
+    }
+
+    int batchedSize = batchedRows.size();
+    // if no sql is batched, just return
+    if (batchedSize == 0) {
+      return new int[0];
+    }
+
+    log.fine(batchedSize + " records are going to be uploaded to table " + tableBatchInsertTo
+             + " in batch");
+
+    int[] updateCounts = new int[batchedSize];
+    Arrays.fill(updateCounts, -1);
+
+    try {
+      TableTunnel.UploadSession session;
+      TableTunnel tunnel = new TableTunnel(getConnection().getOdps());
+
+      if (tableBatchInsertTo.contains(".")) {
+        String[] splited = tableBatchInsertTo.split("\\.");
+        session = tunnel.createUploadSession(splited[0], splited[1]);
+      } else {
+        String project_name = getConnection().getOdps().getDefaultProject();
+        session = tunnel.createUploadSession(project_name, tableBatchInsertTo);
+      }
+      log.info("create upload session id=" + session.getId());
+
+      TableSchema schema = session.getSchema();
+      Record record = session.newRecord();
+      int colNum = schema.getColumns().size();
+      int valNum = batchedRows.get(0).length;
+      if (valNum != colNum) {
+        throw new SQLException(
+            "the table has " + colNum + " columns, but insert " + valNum + " values");
+      }
+
+      long startTime = System.currentTimeMillis();
+      TunnelRecordWriter recordWriter = (TunnelRecordWriter) session.openRecordWriter(0, true);
+      for (int i = 0; i < batchedSize; i++) {
+        Object[] row = batchedRows.get(i);
+        for (int col = 0; col < colNum; col++) {
```

Comment:
yep, trade correctness for performance. Currently, CPU time is not the bottleneck.